keps/sig-node/2053-downward-api-hugepages/README.md

```diff
@@ -0,0 +1,242 @@
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+This KEP exposes hugepages in the downward API.
+
+## Motivation
+
+Pods are unable to know their hugepage request or limits via the downward API.  HugePages
+are a natively supported resource in Kubernetes and should be visible in downward API
+consistent with other resources like cpu, memory, ephemeral-storage.
+
+### Goals
+
+- Add support for hugepage requests and limits for all page sizes in downward API
+
+### Non-Goals
+
+- Change any other aspect of hugepage support
+
+## Proposal
+
+Define a new feature gate: `DownwardAPIHugePages`.
+
+If enabled, the `kube-apiserver` will allow pod specifications to make use
+of hugepages in downward API when the feature gate is enabled.  The `kubelet`
+will add support for hugepages in the downward API independent of the feature
+gate.
+
+### Risks and Mitigations
+
+The primary risk for this proposal is that it loosens validation for Pods.
+
+The mitigation proposed is as follows:
+
+- Add support for the new fields in `kubelet` by default.  This is considered
+low risk as the code is inert when pods do not use the tokens, and the subsystem
+in the kubelet is localized.
+- The `kube-apiserver` will have the feature gate disabled by default for 2
+releases until we know all supported skew scenarios result in all kubelets having
+the supported code present.
+
+When the gate is enabled, the `kube-apiserver` will permit the newly allowed
+values in all creation and update scenarios.  When the gate is disabled, the
+new values are permitted only in updates of objects which already contain
+the new values.  Use in creation of in updates of objects which do not
+already use the new values will fail validation.
+
+## Design Details
+
+Add support for `requests.hugepages-<pagesize>` and `limits.hugepages-<pagesize>`
```

Comment:
will validation ensure that only pagesizes actually present in requests/limits are allowed? if not, what value is used if a pagesize that is not set as a request or limit is specified as a downward API value?

Reply:
value projected in pod will always match value presented in the cgroup (which absent an explicit request is 0) because hugepages are not subject to overcommit like memory.  validation will not require a value present in request/limit as i anticipate pod templates will use the downward API form in order to know if a resource request was made or not for specific page sizes as the same workload is applied to nodes with different page sizes, or have some kustomize template that omits the request when testing, but the workload would still want to know it has 0 pages at that size..

Reply:
> i anticipate pod templates will use the downward API form in order to know if a resource request was made or not for specific page sizes as the same workload is applied to nodes with different page sizes

Are supported page sizes well-known? Would a pod template have to enumerate a lot of pagesizes to be aware of all of the possible values? How would an injected container discover hugepage requests/limits via this mechanism without first-hand knowledge of the pod template it was injected into? All of the existing downwardAPI resource keys are fixed (e.g. memory, cpu, ephemeral-storage), so it's easy to discovery resource configurations.

Reply:
@liggitt the list of pagesizes varies by architecture.  there are a practical set of pagesizes supported by things like runc.  the validation for the downward api field will not vary from the validation for resource requirements today for hugepages.